### PR TITLE
MINOR: Move TopicIdPartition class to server-common

### PR DESCRIPTION
**Pull request by rreddy-22**
Moving the TopicIdPartition class from the metadata controller package to the server-common package so that it can be easily used by other classes. 
Immediately required by the server side assignors for the new consumer group protocol aka KIP-848.

### CODE REVIEW

File: server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java [hunk 1]
```diff
@@ -14,43 +14,62 @@
...
 import org.apache.kafka.common.Uuid;
 
-final class TopicIdPartition {
+/**
+ * Represents a partition using its unique topic Id and partition number.
+ */
+public final class TopicIdPartition {
     private final Uuid topicId;
-    private final int partitionId;
+    private final Integer partitionId;
```
Comment (dajac) [inline server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java]: Why changing to an Integer? An int seems more appropriate.

File: server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java [hunk 2]
```diff
@@ -14,43 +14,62 @@
...
     public int partitionId() {
         return partitionId;
     }
 
     @Override
     public boolean equals(Object o) {
-        if (!(o instanceof TopicIdPartition)) return false;
-        TopicIdPartition other = (TopicIdPartition) o;
-        return other.topicId.equals(topicId) && other.partitionId == partitionId;
+        if (this == o) {
```
Comment (dajac) [inline server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java]: nit: Should we keep the previous implementation for equals, hashCode and toString? We just want to move the class so we should not change it.

File: server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java [hunk 3]
```diff
@@ -14,43 +14,62 @@
...
     @Override
     public String toString() {
-        return topicId + ":" + partitionId;
+        return "TopicIdPartition(" +
+            "topicId=" + topicId +
+            ", partition=" + partitionId +
+            ')';
     }
-}
+}
```
Comment (dajac) [inline server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java]: Let's add an empty line at the end of the file as it was.

**Review by dajac — COMMENTED** [review]
@rreddy-22 Thanks for the PR. I left a few comments.

File: server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java (hunk 3 above)
Comment (rreddy-22) [reply server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java]: ohno I thought we had to remove the extra line, I remember getting a comment before, I'll change it in all the PR's

File: server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java (hunk 2 above)
Comment (rreddy-22) [reply server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java]: I figured while we're at it we could update the methods according to when we reviewed TopicIdPartition file that I created in the coordinator common class. I could revert it if you feel the old one is fine

File: server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java (hunk 1 above)
Comment (rreddy-22) [reply server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java]: Couldn't use equals and hashCode methods on int

Comment (dajac) [reply server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java]: We should keep an int here. It is possible to implement equals and hashCode with it (as it was before).

File: server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java (hunk 2 above)
Comment (dajac) [reply server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java]: Let's keep it as it was.

**Review by vamossagar12 — APPROVED** [review]
Thanks for this PR @rreddy-22 . LGTM. I had a slightly orthogonal question , I see another class by the same name TopicIDPartition [here](https://github.com/apache/kafka/blob/7ba6d7a0b439cd7be7918b77c6b08425d6b37526/clients/src/main/java/org/apache/kafka/common/TopicIdPartition.java#L25) within the clinets module. Ideally the users of that class should use the class from server-common since that's the intent I believe. We don't need to derail this PR but probably we can file another ticket which does the same. WDYT?

**Comment by dajac** [comment]
Re-triggered jenkins.

File: server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java [hunk 4]
```diff
@@ -54,3 +63,6 @@ public String toString() {
         return topicId + ":" + partitionId;
     }
 }
+
+
+
```
Comment (dajac) [inline server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java]: nit: Could we remove those two empty lines? We just need one at the end of the file.

**Review by dajac — COMMENTED** [review]
@rreddy-22 Thanks for the update. I just left one small nit. We can merge it when addressed.

**Comment by rreddy-22** [comment]
> Thanks for this PR @rreddy-22 . LGTM. I had a slightly orthogonal question , I see another class by the same name TopicIDPartition [here](https://github.com/apache/kafka/blob/7ba6d7a0b439cd7be7918b77c6b08425d6b37526/clients/src/main/java/org/apache/kafka/common/TopicIdPartition.java#L25) within the clinets module. Ideally the users of that class should use the class from server-common since that's the intent I believe. We don't need to derail this PR but probably we can file another ticket which does the same. WDYT?

Hey, so that class has topicId mapped to TopicPartition and here we just use integer partition so we can't reuse this class there. I agree though that two classes with the same name isn't good

**Review by dajac — APPROVED** [review]
LGTM

**Pull request merged**